Opencast-Moodle/moodle-block_opencast
# Keep seriesmapping records when deleting a block (configurable)

This PR fixes #351

### Description
please refer to the related issue #351

### Important to know
- The logical way (current way) is to remove any related records upon deleting a block, therefore you do not need to worry about the rest of cleaning and everything else. With that said, the current way of removing `seresmapping` record upon deleting a block is correct but not enough.
- This feature request will leave orphan records in the database table (which is a NO-GO), that results in a huge and messy table in a long-run that compromises the performance. Therefore, it is designed to be configurable that brings attention to its cons!

### How it works
- In plugin settings under `General settings > Settings for a block instance` a new setting option is introduced `Keep serie mapping after deleting the block` which toggles this feature.
- When activated, it refuses to delete the course-series mapping record from the related table, otherwise, it goes on like before.

### How to test
- Make sure you have the latest plugin versions.
- Have this PR patched.
- Head over to plugin admin setting and enable the above mentioned setting option.
- In a test course with edit mode on:
   - Make sure Opencast Block is added and has (content) a series and videos in it.
   - Delete the Opencast Block.
   - Add the Opencast Block again.
   - Confirm that the series and videos are there and nothing is missing!
- To test the deactivated scenario, just repeat the above steps with the setting off, but make sure your videos and series in that block are temporary and are for the sake of this test!

### What needs to be done in near future
The proper way to do is to have a back-up table that records the deleted courses and series mapping and get the record back when re-adding the block. In this way, you have a separate table and the performance won't be compromised.

**NOTE:** behat test scenario is included
